### DISCUSSION TITLE
How to prevent additional CROSS JOIN when using CTE in TimescaleDB's LOCF function

### DISCUSSION THREAD

**Discussion by mdczaplicki** [sqlalchemy/sqlalchemy#12181] [Usage Questions] [answered]
I'm creating a CTE that I'd like to use as  a `prev` parameter if `LOCF` function from TimescaleDB (https://docs.timescale.com/api/latest/hyperfunctions/gapfilling/time_bucket_gapfill/#locf).
I have a problem that SQLAlchemy generates an additional `CROSS JOIN` between an actual table and that CTE, whereas that CTE shouldn't be present in the `JOIN` section, only in the `SELECT` part.

Here's part of the Python code:

```
previous_day_cte = (
    select(
        [
            func.avg(
                MyTable.some_value
            ).label("prev")
        ]
    )
    .where(MyTable.minute == previous_day_minute)
    .group_by(MyTable.minute)
    .cte("previous_day")
)
rb_cte = (
    select(
        [
            my_table_bucket.label(timestamp_label),
            func.locf(
                func.avg(MyTable.some_value),
                select([previous_day_cte.c.prev]),  # This is `prev` parameter of TimescaleDB `locf`
            ).label("some_value"),
        ]
    )
    .select_from(MyTable)
    .where(MyTable.minute >= start_datetime)
    .where(MyTable.minute < end_datetime)
    .group_by(timestamp_label)
    .cte("rb")
)
```
And here's the query that's being generated. (note the `FROM my_table, (SELECT ...` part of the query).

```
WITH previous_day AS 
(SELECT avg(my_table.some_value) AS prev 
FROM my_table 
WHERE my_table.minute = '2000-10-10T10:09:00+00:00' GROUP BY my_table.minute), 
rb AS 
(SELECT time_bucket_gapfill('1 minute', my_table.minute) AS timestamp, locf(avg(my_table.some_value), (SELECT previous_day_rb.prev 
FROM previous_day_rb)) AS some_value 
FROM my_table, (SELECT previous_day_rb.prev AS prev 
FROM previous_day_rb) 
WHERE my_table.minute >= '2000-10-10T10:10:00+00:00' AND my_table.minute < '2000-10-10T10:15:00+00:00' GROUP BY timestamp), 
```

How can I prevent SQLAlchemy from adding this additional `JOIN`?

I'm running on SQLAlchemy 1.3.24, fairly old, but can't upgrade right now.

**Comment by zzzeek** [comment]
I can't experiment with the example since it's incomplete and it's not clear to what extent CTEs in 1.3 can refer to each other but you'd at least look to see if you can get the correlate feature to work with you here, by adding `.correlate_except(MyTable)` to that second SELECT.  but i dont know if correlate is going to work in the CTE context.      There have been many CTE fixes over the course of 1.4 that you wont have in 1.3.

**Reply by mdczaplicki** [reply]
Here's a reproducible example:
```
from sqlalchemy.ext.declarative import (
    declarative_base,
)
from sqlalchemy import Column, Integer, select, func

Base = declarative_base()

class MyTable(Base):
    __tablename__ = "MyTable"
    a = Column(Integer, primary_key=True)
    some_value = Column(Integer)

cte = select([MyTable.some_value.label("prev")]).select_from(MyTable).cte("cte")

query = select([func.avg(select([cte.c.prev]).select_from(cte)), MyTable.a]).select_from(MyTable)

print(query.compile())
```

`.correlate_except` doesn't seem to have any effect.

**Reply by mdczaplicki** [reply]
If there's no way to do it nicely, how could I achieve it nasty?

This is what I came up with:

```
query1 = select([func.avg(text("SELECT prev FROM cte")), MyTable.a]).select_from(MyTable)
```
but it doesn't include CTE at all.

**Reply by zzzeek** [reply, marked as answer]
thanks for the MCVE.  I can confirm this is fixed in 1.4, however based on the commit that fixed it, it reveals how to make it work in 1.3

here's 1.3:

```
WITH cte AS 
(SELECT "MyTable".some_value AS prev 
FROM "MyTable")
 SELECT avg((SELECT cte.prev 
FROM cte)) AS avg_1, "MyTable".a 
FROM (SELECT cte.prev AS prev 
FROM cte), "MyTable"

```

1.4:

```
WITH cte AS 
(SELECT "MyTable".some_value AS prev 
FROM "MyTable")
 SELECT avg((SELECT cte.prev 
FROM cte)) AS avg_1, "MyTable".a 
FROM "MyTable"

```

but the only issue is that in 1.3 you are not indicating that your `func.avg()` query is a scalar subquery, and 1.3 is seeing it as a FROM clause, so this fixes:

```
query = select(
    [func.avg(select([cte.c.prev]).as_scalar()), MyTable.a]
).select_from(MyTable)
```

**Reply by mdczaplicki** [reply]
Thanks Michael! You're the boss! 😎 ❤️
